### PR TITLE
NuGet Lockfile Improvements and Caching Removal

### PR DESCRIPTION
**Pull request by NickGerleman**
Enforces that lockfiles are updated in CI, so that we do not see lockfile changes during build interfering with cache. Updated related projects to look more like template, and fix non-deterministic lockfiles.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/9093)

### CODE REVIEW

File: .ado/templates/run-windows-with-certificates.yml [hunk 1]
```diff
@@ -23,20 +23,32 @@ parameters:
...
+    parameters:
+      solutionPath: ${{ parameters.workingDirectory }}
+      nugetConfigPath: ${{ parameters.workingDirectory }}/windows/NuGet.config
+      repositoryPath : ${{ parameters.workingDirectory }}/windows/packages
 
   - ${{ if eq(parameters.buildConfiguration, 'Debug') }}:
-    - script: yarn windows --no-packager --no-launch ${{ parameters.deployOption }} --arch ${{ parameters.buildPlatform }} --logging --buildLogDirectory ${{ parameters.buildLogDirectory }}
+    - script: >
+        yarn windows
+        --arch ${{ parameters.buildPlatform }}
```
Comment (JunielKatarn) [inline .ado/templates/run-windows-with-certificates.yml:35]: (nit)

Indent arguments one level beyond command?
i.e.:

```
  yarn windows
    --arch ...
```

Comment (NickGerleman) [reply .ado/templates/run-windows-with-certificates.yml:35]: I was seeing some issues with that, where I needed to deindent :/

**Comment by NickGerleman** [comment]
Playground before: 
image

**Comment by NickGerleman** [comment]
Playground after:
image

**Comment by NickGerleman** [comment]
Universal before:
image

**Comment by NickGerleman** [comment]
Universal after:
image

**Comment by NickGerleman** [comment]
Universal, the job in the critical path, is showing improvements. Notably we reduce physical content downloaded from 2,435.9 MB to 1,078.8 MB, which should hopefully cut down on issues.

**Comment by NickGerleman** [comment]
Some of the universal runs are even faster. 
image

File: .ado/templates/cache-nuget-restore.yml [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    condition: ne(variables.nugetCacheHit, 'false')
+
+  - pwsh: |
+      gci -File -Recurse -Filter *.nupkg $(Agent.TempDirectory)\nupkg | ForEach-Object -Parallel {
+        $outputFile = $_.FullName.Replace('$(Agent.TempDirectory)\nupkg','${{ parameters.globalPackagesFolder }}')
+        Copy-Item $_.FullName _Destionation $outputFile -Recurse
+
+        # 7-zip is orders of magnitude faster than Expand-Archive;
+        $outputDir = $_.Directory.FullName.Replace('$(Agent.TempDirectory)\nupkg','${{ parameters.globalPackagesFolder }}')
+        & $env:ProgramFiles\7-Zip\7z.exe x -y "-o$($outputDir)\" "$($_.FullName)"
```
Comment (dannyvv) [inline .ado/templates/cache-nuget-restore.yml:65]: Not sure if 7zip does the 'right' thing...
Nuget is a zip-file but it is a zip-file with some semantics and path translations that are supposed to happen.

if the .zip file contains `lib/netcore50/my%20lib.dll`
The extracted file should be `lib/netcore50/my lib.dll`. If 7zip does the right thing, then no worries, else we might have to use a different library and manually extract the package, or actually use nuget.exe for that...

Comment (dannyvv) [reply .ado/templates/cache-nuget-restore.yml:65]: and you take care of that later... Please ignore my ignorance :P

**Comment by dannyvv** [comment]
After being horrified by the need of all this stuff, an old idea I forgot about finally popped back in my head.

Since this is mostly to deal with nuget.org reliability, why don't we take a slightly different approach...

As part of creating the image we:
* set the nuget cache environment to the `c:\cache\nuget`
* clone the repo to `c:\imagesetup\react-native-windows`
* run nuget restore
* remove `c:\imagesetup\react-native-windows`

And of course the same for:
* npm/npx
* yarn 
* midgard-yarn

although we'll need to add a flag ot all yarn and midgard-yarn to prioritize local cache over network....

I tried to do this before the 1ES hosted images when creating a private azure scale set pool. But didn't see a convenient setup of managed images.

**Comment by NickGerleman** [comment]
> After being horrified by the need of all this stuff, an old idea I forgot about finally popped back in my head.
> 
> Since this is mostly to deal with nuget.org reliability, why don't we take a slightly different approach...
> 
> As part of creating the image we:
> 
> * set the nuget cache environment to the `c:\cache\nuget`
> * clone the repo to `c:\imagesetup\react-native-windows`
> * run nuget restore
> * remove `c:\imagesetup\react-native-windows`
> 
> And of course the same for:
> 
> * npm/npx
> * yarn
> * midgard-yarn
> 
> although we'll need to add a flag ot all yarn and midgard-yarn to prioritize local cache over network....
> 
> I tried to do this before the 1ES hosted images when creating a private azure scale set pool. But didn't see a convenient setup of managed images.

I looked at caching during image creation. Provisioning scripts are run after image creation, so I think we would need to add our own artifact...

**Comment by NickGerleman** [comment]
See https://github.com/microsoft/react-native-windows/issues/8990 for reference on the above.

**Comment by dannyvv** [comment]
> I looked at caching during image creation. Provisioning scripts are run after image creation, so I think we would need to add our own artifact...

Yes, the way to do it would be to submit an 'action' that allows for custom script during image creation...

**Comment by NickGerleman** [comment]
A potential downside to that is longer provisioning time from a larger image, or more cache misses on older branches. I imagine it would work very well in the common case though. 

It would be worth experimenting with, but I think with this change, I've gotten blob storage caching into a relatively good place (for perf + reliability compared to NuGet.org). Though the ultimate test is to live with it for a bit and observe to see if there are runs with reliability issues or unexpected latencies.

File: .ado/templates/run-windows-with-certificates.yml [hunk 3]
```diff
@@ -21,22 +21,33 @@ parameters:
...
+        --buildLogDirectory ${{ parameters.buildLogDirectory }}
+        --msbuildprops RestoreLockedMode=${{ parameters.restoreLockedMode }}
+        ${{ parameters.deployOption }}
       displayName: run-windows (Debug)
       workingDirectory: ${{ parameters.workingDirectory }}
 
   - ${{ if and(eq(parameters.buildConfiguration, 'Release'), eq(parameters.buildEnvironment, 'PullRequest')) }}:
-    - script: yarn windows --no-packager --no-launch ${{ parameters.deployOption }} --arch ${{ parameters.buildPlatform }} --logging --buildLogDirectory ${{ parameters.buildLogDirectory }} --release
+    - script: >
+        yarn windows
```
Comment (dannyvv) [inline .ado/templates/run-windows-with-certificates.yml:43]: just checking if you are intentionally dropping `--no-packager`

**Review by dannyvv — APPROVED** [review]

File: .ado/templates/run-windows-with-certificates.yml (hunk 3 above)
Comment (NickGerleman) [reply .ado/templates/run-windows-with-certificates.yml:43]: Yep. `--no-packager` has been implied by `--no-launch` for a few versions.

**Pull request merged**